### ISSUE TITLE
[MGPG-136] "gpg: signing failed: Bad passphrase" on GitHub Windows runners

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-gpg-plugin#172]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** opened **[MGPG-136](https://issues.apache.org/jira/browse/MGPG-136?redirect=false)** and commented

Version 3.2.0 - 3.2.6 fail with "gpg: signing failed: Bad passphrase" on GitHub Windows Runners.

Using Linux or downgrading to version 3.1.0 seems to work fine.

It's not clear what changed in these newer versions but https://github.com/apache/maven-gpg-plugin/blob/3a31714e9cbdde86a6b12f1ef05d5c60252fef4a/src/main/java/org/apache/maven/plugins/gpg/AbstractGpgMojo.java#L368 looks like a highly supicious way of reading environment variables. It's possible that this approach does not work properly under Powershell.

If I echo "$env:MAVEN_GPG_PASSPHRASE" on the GitHub Windows runner, I get back the correct value, so it sounds like there is a bug in the way that this plugin is choosing to read the environment variable.

What makes this a bit tricky, however, is that the environment variable is called $MAVEN_GPG_PASSPHRASE on Linux, %MAVEN_GPG_PASSPHRASE% on cmd.exe and $env:MAVEN_GPG_PASSPHRASE on Powershell. GitHub is using the Powershell variant.

Can you guys please look into this?

---

**Affects:** 3.2.0, 3.2.1, 3.2.2, 3.2.3, 3.2.4, 3.2.5, 3.2.6

**Issue Links:**
- [MGPG-145](https://issues.apache.org/jira/browse/MGPG-145) (IT suite) IT did not catch issue on Windows OS

- [MGPG-99](https://issues.apache.org/jira/browse/MGPG-99) Passcode byte array provided to gpg executable on stdin is not terminated

**Remote Links:**
- [GitHub Pull Request #120
](https://github.com/apache/maven-gpg-plugin/pull/120)

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

GPG plugin does not reach directly for env variable, Maven does it on boot, and GPG plugin just gets that from session.
To dump what Maven "see", you can use command like this:

```
$ mvn eu.maveniverse.maven.plugins:toolbox:0.1.33:dump -Dverbose
```

(note: 0.1.33 is just released as I had some fixes in there, it may need few minutes for Central sync).

This command dumps all that is present in repository system session, the object GPG plugin uses as well to get values.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas) I used the command you mentioned and it confirms that "env.MAVEN_GPG_PASSPHRASE" contains a secret value (the value shows up as ***).

I am the second person to report this bug (see https://github.com/actions/setup-java/issues/629).

Can you please debug this problem on your end to figure out why the passphrase is being rejected?

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas) Please see https://github.com/actions/setup-java/issues/629#issuecomment-2333230328 and let us know if any other ideas come to mind. Thank you in advance.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Howdy Gili, thanks for reaching back with new info, and sorry for a bit late response.
You pointed at line that is "highly suspicious way of reading environment variables", so let me start there:
Maven at "boot" collects Java System Properties  and env variables into Maven System Properties (not to be confused with Java System Properties, that are ONE SOURCE of these). Same things happens in Toolbox (and there you confirmed that env.MAVEN_GPG_PASSPHRASE contains a secret value). And don't be confused, the "env." prefix is nothing to do with PowerShell or whatever, it is really just Maven:
See https://github.com/apache/maven/blob/45201347c417896b57159221130333f5fa4bbfb6/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java#L1485
And https://github.com/apache/maven/blob/45f9b81b4a8451a75864ef1c861c5bb201a54790/maven-core/src/main/java/org/apache/maven/properties/internal/EnvironmentUtils.java#L34

In short, Maven system properties contain Java System Properties (as is) AND env variables where key are prefixed with "env.". Hence the way of "reading".

Also, just like between Linux/MacOS/Windows, these differences _should_ be handled by Java, am pretty sceptic that in Java you should key env variable differently based on cmd/powershell... or if this is the case on Windows, than this platform is eeek.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

But I found a "sus" place myself, but for this to confirm, I'd need some answers: do your passphrases contains some character that is beyond 7bit ASCII?

https://github.com/apache/maven-gpg-plugin/blob/master/src/main/java/org/apache/maven/plugins/gpg/GpgSigner.java#L115-L120

This place smells like it lacks some encoding, as "usually" Mac and Linuxes are UTF8, but AFAIK Windows is stubborn on that `chcp 65001`?

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

In 3.1.0 the process input creation was like this:
https://github.com/apache/maven-gpg-plugin/blob/maven-gpg-plugin-3.1.0/src/main/java/org/apache/maven/plugins/gpg/GpgSigner.java#L105-L106

So it seems Windows is getting passphrase+newline? Hence GPG declares it as "bad passphrase"? Again, am unsure how Windows behave in this respect...

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Help!

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

So two issues:
* possible encoding issues here: Java def encoding is used to fill in ByteArrayInputStream, that is later as-is used to feed process. On Linux/mac the java encoding is usually == console encoding == utf8, while this AFAIK is not true for window? https://github.com/apache/maven-gpg-plugin/blob/master/src/main/java/org/apache/maven/plugins/gpg/GpgSigner.java#L115-L120
* possible passphrase corruption as well, by adding System.lineSeparator() ?

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc) **wait a sec...** you say "I used the command you mentioned and it confirms that "env.MAVEN_GPG_PASSPHRASE" contains a secret value (the value shows up as ***)."?
That is weird, as Toolbox (the command I said to run), does not hide sensitive values as it have no idea which one is sensitive!

I just tested it:

```
$ MAVEN_GPG_PASSPHRASE=bigSecretNobodyCanSee mvn eu.maveniverse.maven.plugins:toolbox:0.3.2:dump -Dverbosity=CHATTER
...
[INFO]                          env.MAVEN_CMD_LINE_ARGS= eu.maveniverse.maven.plugins:toolbox:0.3.2:dump -Dverbosity=CHATTER
[INFO]                          env.MAVEN_GPG_PASSPHRASE=bigSecretNobodyCanSee
[INFO]                          env.MAVEN_HOME=/home/cstamas/.sdkman/candidates/maven/current
...
```

This means your passphrase is "****"

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Hi Tamas,

Thanks for the follow-up... The reason that I see *** in place of the secret is that I'm running the command under a GitHub Action. It is GitHub that is hiding the output, not toolbox itself. With respect to my passphrase, the unencrypted version is purely alphanumeric so no, I don't expect any characters outside the 7bit range.

Let me know if there is anything else I can do to help tracking down this issue. If you release a snapshot version or something I can try running it under Github again.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc) I created this https://github.com/ralfkonrad/test-mvn-gpg-plugin/pull/27 but needs auth. You can try same if you can.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc) seems it was rogue CR!
See https://github.com/ralfkonrad/test-mvn-gpg-plugin/pull/28

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Question: why does this even need to be a user-configurable option? To me, this should be an internal implementation detail that is not configurable by end-users.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

You are right, this is really internal detail.
Still, having it as parameter is okay, even if nobody will have to use it. We can schedule removal of this parameter if you strongly believe this should be not exposed.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Your call, though it's harder to remove something once it's been published. Either way, thank you for chasing this issue down!
